### PR TITLE
Implement missing endpoints

### PR DESCRIPTION
**Pull request by magnusbechwind**
## Purpose

Implement the missing endpoints: `GetBakerEarliestWinTime`, `GetBakersRewardPeriod`, `GetBlockCertificates`, `GetWinningBakersEpoch`,  and `GetFirstBlockEpoch`.

## Changes

Updated the grpc api. Added `EpochRequestBlockHash` and `EpochRequestRelativeEpoch` to `types.go` and implemented the endpoints mentioned above.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: v2/tests/api_test.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ import (
 	"context"
 	"testing"
 
-	"github.com/Concordium/concordium-go-sdk/v2"
+	v2 "github.com/Concordium/concordium-go-sdk/v2"
```
Comment (abizjak) [inline v2/tests/api_test.go:7]: Why is this needed?

File: v2/types.go [hunk 2]
```diff
@@ -995,3 +995,59 @@ func ConvertBlockItems(input []*pb.BlockItem) []*BlockItem {
...
+	isEpochRequest()
+}
+
+// Query for the epoch of a specified block.
+type EpochRequestBlockHash struct {
+	// The block to query at.
+	BlockHash isBlockHashInput
+}
+
+func (EpochRequestBlockHash) isEpochRequest() {}
```
Comment (abizjak) [inline v2/types.go:1015]: I don't understand what this function  is supposed to do.

File: v2/types.go [hunk 3]
```diff
@@ -995,3 +995,59 @@ func ConvertBlockItems(input []*pb.BlockItem) []*BlockItem {
...
 }
+
+// A chain epoch.
+type Epoch struct {
+	Value uint64
+}
+
+// Input to queries which take an epoch as a parameter.
+type isEpochRequest interface {
+	isEpochRequest()
```
Comment (abizjak) [inline v2/types.go:1006]: It's odd to have this function. Why do we need this interface with this function that does nothing?

File: v2/getfirstblockepoch.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//
+// The following errors are possible:
+//   - `NOT_FOUND` if the query specifies an unknown block.
+//   - `UNAVAILABLE` if the query is for an epoch that is not finalized in the current genesis index, or is for a future genesis index.
+//   - `INVALID_ARGUMENT` if the query is for an epoch with no finalized blocks for a past genesis index.
+//   - `INVALID_ARGUMENT` if the input `EpochRequest` is malformed.
+//   - `UNIMPLEMENTED` if the endpoint is disabled on the node.
+//
+// This endpoint is only supported for protocol version 6 and onwards.
+func (c *Client) GetFirstBlockEpoch(ctx context.Context, req isEpochRequest) (_ *pb.BlockHash, err error) {
```
Comment (abizjak) [inline v2/getfirstblockepoch.go:19]: Why are we not using `v2.BlockHash` as the return value? That ensures the value is 32-bytes.

File: v2/getwinningbakersepoch.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// The following error cases are possible:
+//   - `NOT_FOUND` if the query specifies an unknown block.
+//   - `UNAVAILABLE` if the query is for an epoch that is not finalized in the current genesis index, or is for a future genesis index.
+//   - `INVALID_ARGUMENT` if the query is for an epoch that is not finalized for a past genesis index.
+//   - `INVALID_ARGUMENT` if the query is for a genesis index at consensus version 0.
+//   - `INVALID_ARGUMENT` if the input `EpochRequest` is malformed.
+//   - `UNIMPLEMENTED` if the endpoint is disabled on the node.
+//
+// This endpoint is only supported for protocol version 6 and onwards.
+func (c *Client) GetWinningBakersEpoch(ctx context.Context, req isEpochRequest) (_ pb.Queries_GetWinningBakersEpochClient, err error) {
```
Comment (abizjak) [inline v2/getwinningbakersepoch.go:24]: Could we make this a bit nicer by adding a wrapper around the GetWinningBakersEpochClient that uses the types such as `BakerId` we have already defined.

File: v2/getbakersrewardperiod.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import (
+	"context"
+
+	"github.com/Concordium/concordium-go-sdk/v2/pb"
+)
+
+// GetBakersRewardPeriod retrieves all bakers in the reward period of a block.
+//
+// This endpoint is only supported for protocol version 6 and onwards.
+func (c *Client) GetBakersRewardPeriod(ctx context.Context, req isBlockHashInput) (_ pb.Queries_GetBakersRewardPeriodClient, err error) {
```
Comment (abizjak) [inline v2/getbakersrewardperiod.go:12]: Could we please add a wrapper for GetBakersRewardPeriodClient that uses `BakerId` type wrapper we have defined in types.go?

File: v2/getbakerearliestwintime.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+//     the actual time may be later, and the reported time in subsequent queries may reflect this.)
+//   - At the end of an epoch (or if the baker is not projected to bake before the end of the epoch)
+//     the earliest win time for a (current) baker will be projected as the start of the next epoch.
+//     This is because the seed for the leader election is updated at the epoch boundary,
+//     and so the winners cannot be predicted beyond that.
+//
+// Note that in some circumstances the returned timestamp can be in the past, especially at the end of an epoch.
+//
+// This endpoint is only supported for protocol version 6 and onwards.
+func (c *Client) GetBakerEarliestWinTime(ctx context.Context, req *pb.BakerId) (_ *pb.Timestamp, err error) {
```
Comment (abizjak) [inline v2/getbakerearliestwintime.go:23]: Please map into the Timestamp type we have defined in `types.go`

**Review by abizjak — COMMENTED** [review]
I though you're going to add also add a submodule reference for the grpc-api module.

I don't see that here.

**Comment by abizjak** [comment]
Also, can you please add some basic examples/tests for the new endpoints.

File: v2/tests/api_test.go (hunk 1 above)
Comment (magnusbechwind) [reply v2/tests/api_test.go:7]: This is added automatically every time i save.

File: v2/types.go (hunk 3 above)
Comment (magnusbechwind) [reply v2/types.go:1006]: Apparently this is how one does interfaces in go, this is essentially a tagging interface. You don't explicictly implement an interface, you just implement every method the interface declares.. Something similar is done for `BlockHashInput`.

File: v2/types.go (hunk 2 above)
Comment (magnusbechwind) [reply v2/types.go:1015]: It implements the `isEpochRequest` interface. See above comment for explanation of why it is done like this.

**Review by abizjak — APPROVED** [review]
